### ISSUE TITLE
3D model deforms when playing animations in Godot 4

### ISSUE THREAD

**Issue by Miartsix** [godotengine/godot#72998]
### Godot version

4.0.rc1

### System information

Windows 10, Vulkan, NVIDIA GeForce GTX 1060 6GB driver version: 27.21.14.5751

### Issue description

When opening a glb. file exported from Blender, the animation works correctly on multiple versions of Godot 3, but deforms in Godot 4. 
I tried changing some exporting settings in Blender and import settings in Godot, but nothing helped. 
![issue_2](https://user-images.githubusercontent.com/91782210/217948411-8043b4f8-4fcc-42f1-ae7d-7db97e1328a4.png)
![issue_1](https://user-images.githubusercontent.com/91782210/217948415-054005d8-ca7b-45fb-b532-5120ebe3a00e.png)

### Steps to reproduce

1. Open the model as a new inherited scene
2. Play the animation

### Minimal reproduction project

[Issue_project.zip](https://github.com/godotengine/godot/files/10701889/Issue_project.zip)

**Comment by clayjohn** [comment]
This is an issue with our current Auto-LOD generator. The auto LODs are being used much too early and they quality is not good for animated meshes. 

You can work around it by disabling LOD generation in the import menu or increasing the LOD bias on the MeshInstance3D

Part of https://github.com/godotengine/godot/issues/57416

**Comment by fire** [comment]
Working on fixing this with https://github.com/godotengine/godot/pull/73734

**Comment by fire** [comment]
I tested in the latest, I think the AABB changes in Godot Engine 4.1 may have fixed.

**Comment by clayjohn** [comment]
> I tested in the latest, I think the AABB changes in Godot Engine 4.1 may have fixed.

Is it fixed, or does it just not appear until the camera is further away than before?

**Comment by ydeltastar** [comment]
I run into this issue a lot while working with animated characters. Bone structure looks fine in the editor but some parts of the mesh is deformed.

It only happens when the model is an Editable Children. If I uncheck Editable Children, it keeps deformed in the editor but works in runtime. If I restart the editor, it is fixed in the editor too. I can reproduce the same behavior in the minimal reproduction project here.
Looking at the git diff of the scene, it saves bones rotations when it's an editable children causing the issue and when you disable it it removes these rotations from the scene file fixing the issue. I guess it is ignoring AnimationPlayer's reset_on_save feature.

**Comment by fire** [comment]
If it's a caching reloading but there are some fixes in that direction. https://github.com/godotengine/godot/pull/82471

**Comment by ydeltastar** [comment]
I tested with a build of #82471 but it doesn't have any effect on the issue.

Disabling LODs generation fixed it in 4.1.2 so the saved bone rotations might be irrelevant. It should still reset on save thought since it creates unnecessary diffs but that sounds like a case for a different issue.

**Comment by pewcworrell** [comment]
I found this while looking into `reset_on_save` being broken. I have a workaround in place in case anybody needs it. I'd imagine solving this issue would fix `reset_on_save` though.
```gdscript
@tool
class_name DummyAnimationPlayer extends AnimationPlayer

@onready var _was_active = active


func _notification(what: int) -> void:
	if !Engine.is_editor_hint():
		return

	match what:
		NOTIFICATION_EDITOR_PRE_SAVE:
			if reset_on_save:
				_was_active = active
				stop()
		NOTIFICATION_EDITOR_POST_SAVE:
			if reset_on_save:
				if _was_active:
					play()
					
```

**Comment by pewcworrell** [comment]
I stumbled upon this issue while investigating why `reset_on_save` wasn't working. Our `Skeleton3D` ends up with changes to all of its animated bones each save. I imagine solving this issue will fix the bone rotation issues I'm seeing. In the mean time, I have a workaround. Just attach this script to your `AnimationPlayer`:
```gdscript
@tool
class_name DummyAnimationPlayer extends AnimationPlayer

@onready var _was_active = active


func _notification(what: int) -> void:
	if !Engine.is_editor_hint():
		return

	match what:
		NOTIFICATION_EDITOR_PRE_SAVE:
			if reset_on_save:
				_was_active = active
				stop()
		NOTIFICATION_EDITOR_POST_SAVE:
			if reset_on_save:
				if _was_active:
					play()
```

**Comment by pewcworrell** [comment]
I've spent quite a bit of time looking at this. I've found that I can reset the position reliably for `AnimationPlayer`s by getting the current animation, calling `Stop`, and then calling `Play` again. For animation trees, I set their `active` to `false`, then set it back to `true`. However, the position is updated immediately _**AFTER**_ saving. So that's useless unless I can figure out how to do it immediately **_before_**, and `NOTIFICATION_EDITOR_PRE_SAVE` isn't working.

```gdscript
func _process_animation_nodes() -> void:
	for scene_root: Node in open_scenes.values():
		var animation_players := scene_root.find_children("*", "AnimationPlayer", true, false)
		var animation_trees := scene_root.find_children("*", "AnimationTree", true, false)
		for animation_player: AnimationPlayer in animation_players:
			_flip_flop_to_reset_animation_player(animation_player)
		for animation_tree: AnimationTree in animation_trees:
			_flip_flop_to_reset_animation_tree(animation_tree)


func _flip_flop_to_reset_animation_player(animation_player: AnimationPlayer) -> void:
	if not animation_player.active:
		return

	var current_animation := animation_player.current_animation
	if current_animation.is_empty():
		var animation_library_list := animation_player.get_animation_library_list()
		if not animation_library_list.is_empty():
			if animation_player.has_animation_library(animation_library_list[0]):
				var lib = animation_player.get_animation_library(animation_library_list[0])
				if lib and lib.get_animation_list().size() > 0:
					if animation_player.has_animation(current_animation):
						current_animation = lib.get_animation_list()[0]
						print(current_animation)
	animation_player.stop()
	if animation_player.active and current_animation != "":
		animation_player.play(current_animation)


func _flip_flop_to_reset_animation_tree(animation_tree: AnimationTree) -> void:
	var was_active = animation_tree.active
	animation_tree.active = false
	animation_tree.active = was_active
```

**Comment by pewcworrell** [comment]
Some of this may not be needed, or may not work perfectly, but for our project it prevents the bones from being different after every save.

**Comment by ydeltastar** [comment]
> Our `Skeleton3D` ends up with changes to all of its animated bones each save.

You need to create a `RESET` track, which will be used as a reference for `reset_on_save`. There is an option in the advanced import settings, added in 4.3 I think, `Animation > Import Rest as Reset` which generates it automatically.

**Comment by pewcworrell** [comment]
I'm almost upset that's an option after how long I spent on this all for the answer to exist on the Skeleton, not the animations 😅. Using the "Import Rest as Reset" while importing the Skeleton's source FBX does work. I still feel like this should be documented better, and maybe a helper method on the AnimationMixer that can trigger a reset would help. Also adding a reference to this in `reset_on_save`, as well as stating in what cases it will and won't work.

For our project, I replaced all `import_rest_as_RESET=false` with `import_rest_as_RESET=true`. I also modified our engine to default this to true. This begs the question though, why is this setting defaulted to false? Is it file size concerns?

**Comment by pewcworrell** [comment]
Ah, this might not work so great. The animations are separate from the skeleton in our files, so we would need to define the RESET.

**Comment by pewcworrell** [comment]
AH. I spoke too soon. My animations weren't animating which is why it appeared fine at first. I have the RESET animation defined, but all transforms are still changing.

**Comment by pewcworrell** [comment]
Judging by your previous comment, perhaps it's the editable children that is the issue?

**Comment by fire** [comment]
As far as I know mesh optimizer works well for MeshInstances3D without skeletons, however in a smaller amount of cases because of the low geometry of the original LOD, when we apply the LODmetric it sometimes breaks animations.

I suggest this issue is a duplicate of https://github.com/godotengine/godot/issues/84479

**Comment by pewcworrell** [comment]
Yeah you're probably right, I think what I'm going on about is a separate issue since I'm using .fbx and not .glb.

**Comment by fire** [comment]
The lod system applies for all scene importers, like not the obj mesh importer.
